stock_storage_type/models/stock_location.py

```diff
@@ -0,0 +1,653 @@
+    @api.depends(
+        "quant_ids.quantity",
+        "out_move_line_ids.qty_done",
+        "in_move_ids",
+        "in_move_line_ids",
+        "only_empty",
+    )
+    def _compute_location_is_empty(self):
+        for rec in self:
+            # No restriction should apply on customer/supplier/...
+            # locations and we don't need to compute is empty
+            # if there is no limit on the location
+            if not rec._should_compute_location_is_empty():
+                # avoid write if not required
+                if not rec.location_is_empty:
+                    rec.location_is_empty = True
+                continue
+            # we do want to keep a write here even if the value is the same
+            # to enforce concurrent transaction safety: 2 moves taking
+            # quantities in a location have to be executed sequentially
+            # or the location could remain "not empty"
+            if (
+                sum(rec.quant_ids.mapped("quantity"))
+                - sum(rec.out_move_line_ids.mapped("qty_done"))
+                > 0
+                or rec.in_move_ids
+                or rec.in_move_line_ids
+            ):
+                rec.location_is_empty = False
+            else:
+                rec.location_is_empty = True
+
+    @api.depends(
+        "storage_category_id.capacity_ids",
+        "location_id",
+        "location_id.allowed_location_capacity_ids",
+    )
+    def _compute_allowed_location_capacity_ids(self):
+        for location in self:
+            if location.storage_category_id.capacity_ids:
+                location.allowed_location_capacity_ids = [
+                    (6, 0, location.storage_category_id.capacity_ids.ids)
+                ]
+            else:
+                parent = location.location_id
+                location.allowed_location_capacity_ids = [
+                    (6, 0, parent.storage_category_id.capacity_ids.ids)
+                ]
+
+    # method provided by "stock_putaway_hook"
+    def _putaway_strategy_finalizer(
+        self,
+        putaway_location,
+        product,
+        quantity=0,
+        package=None,
+        packaging=None,
+        additional_qty=None,
+    ):
+        putaway_location = super()._putaway_strategy_finalizer(
+            putaway_location, product, quantity, package, packaging, additional_qty
+        )
+        if package:
+            quants = package.quant_ids
+            product = package.single_product_id or product
+        else:
+            quants = self.env["stock.quant"].browse()
+        return self._get_pack_putaway_strategy(putaway_location, quants, product)
+
+    def _get_pack_putaway_strategy(self, putaway_location, quants, product):
+        package_type = False
+        if quants:
+            package_type = quants.package_id.package_type_id
+            _logger.debug(
+                "Computing putaway for pack %s (%s)"
+                % (quants.package_id, quants.package_id)
+            )
```

Comment:
@rousseldenis When no quant is provided you should look on the product to get the `package_type_id` . see https://github.com/acsone/wms/commit/48454dcb6731fc2c4bc805a882a82713e1a93ad4

Reply:
@lmignon This has been solved there https://github.com/acsone/wms/commit/48454dcb6731fc2c4bc805a882a82713e1a93ad4